src/Stack/Upgrade.hs

```diff
@@ -29,16 +48,182 @@ import           Stack.Types.Config
         -> Maybe String -- ^ git hash at time of building, if known
+        -> UpgradeOpts
         -> m ()
-upgrade gConfigMonoid gitRepo mresolver builtHash =
+upgrade gConfigMonoid mresolver builtHash uo@(UpgradeOpts mbo mso) =
+    case (mbo, mso) of
+        -- FIXME It would be far nicer to capture this case in the
+        -- options parser itself so we get better error messages, but
+        -- I can't think of a way to make it happen.
+        (Nothing, Nothing) -> error "You must allow either binary or source upgrade paths"
+        (Just bo, Nothing) -> binary bo
+        (Nothing, Just so) -> source so
+        (Just bo, Just so) -> binary bo `catchAny` \e -> do
+            $logWarn "Exception occured when trying to perform binary upgrade:"
+            $logWarn $ T.pack $ show e
+            $logWarn "Falling back to source upgrade"
+
+            source so
+  where
+    binary bo = binaryUpgrade bo
+    source so = sourceUpgrade gConfigMonoid mresolver builtHash so
+
+binaryUpgrade
+  :: (StackM env m, HasConfig env)
+  => BinaryOpts
+  -> m ()
+binaryUpgrade (BinaryOpts platform) = do
+    $logInfo $ "Querying for archive location for platform: " <> platform
+    archiveURL <- liftIO $ findArchive platform
+
+    $logInfo $ "Downloading from: " <> archiveURL
+
+    config <- askConfig
+    let destFile = toFilePath (configLocalBin config </> $(mkRelFile "stack"))
+#if WINDOWS
+                        FP.<.> "exe"
+#endif
+        tmpFile = destFile FP.<.> "tmp"
+
+    liftIO $ do
+      case () of
+        ()
+          | ".tar.gz" `T.isSuffixOf` archiveURL -> handleTarball archiveURL tmpFile
+          | ".zip" `T.isSuffixOf` archiveURL -> error "FIXME: Handle zip files"
+          | otherwise -> error $ "Unknown archive format for Stack archive: " ++ T.unpack archiveURL
+
+    $logInfo "Download complete, testing executable"
+
+    liftIO $ do
+      absTmpFile <- IO.canonicalizePath tmpFile
+
+#if !WINDOWS
+      setFileMode absTmpFile 0o755
+#endif
+
+      -- Sanity check!
+      ec <- rawSystem absTmpFile ["--version"]
+
+      unless (ec == ExitSuccess)
+              $ error $ "Non-success exit code from running newly downloaded executable"
+
+      IO.renameFile tmpFile destFile
+
+    $logInfo $ T.pack $ "New stack executable available at " ++ destFile
+  where
+    setUserAgent :: Request -> Request
+    setUserAgent = setRequestHeader "User-Agent" ["Haskell Stack Upgrade"]
+
+    findArchive :: T.Text -> IO T.Text
+    findArchive pattern = do
+        val <- getResponseBody <$> httpJSON req
+        let ret = do
+                Object top <- return val
+                Array assets <- HashMap.lookup "assets" top
+                getFirst $ fold $ fmap (First . findMatch pattern') assets
+        case ret of
+            Just url -> return url
+            Nothing -> error $ "Could not find Stack archive for platform: " ++ T.unpack pattern
+      where
+        -- FIXME make the Github repo configurable?
```

Comment:
It could, but why add the dependency on another external service that could fail? The logic I've put in here is essentially copy-pasted from stackage-server. (This _might_ argue for making a shared library for easier maintainability going forward, but that's a general Stack maintenance question we should be discussing.)